File: printful.py
"""
Printful API client — SECOND fulfilment provider (apparel).

Read-only catalog use for now (browse what's available / trending). Ordering
comes later and needs a Printful *store* (X-PF-Store-Id) — the account behind
this key currently has none, so order calls are intentionally not wired yet.

Auth: `Authorization: Bearer <PRINTFUL_API_KEY>` (read from a git-ignored .env).
Base: https://api.printful.com. Zero third-party deps (urllib).

Printful identifies a printable variant by an INTEGER catalog variant id
(e.g. 4012) — completely unlike Gelato's long token UIDs. providers.py uses
that difference as a hard guard so the two systems can never be confused.
"""
from __future__ import annotations

import json
import os
import ssl
import urllib.error
import urllib.request
from pathlib import Path

try:
    import certifi
    _SSL_CTX = ssl.create_default_context(cafile=certifi.where())
except Exception:
    _SSL_CTX = ssl.create_default_context()

BASE = "https://api.printful.com"
_ENV = Path(__file__).resolve().parent / ".env"


def _load_env() -> None:
    if not _ENV.exists():
        return
    for line in _ENV.read_text().splitlines():
        line = line.strip()
        if not line or line.startswith("#") or "=" not in line:
            continue
        k, v = line.split("=", 1)
        os.environ.setdefault(k.strip(), v.strip().strip('"').strip("'"))


_load_env()


class PrintfulError(RuntimeError):
    pass


def api_key() -> str:
    return os.environ.get("PRINTFUL_API_KEY", "").strip()


def has_key() -> bool:
    return bool(api_key())


def _request(method: str, path: str, body: dict | None = None,
             store_id: str | None = None) -> dict:
    if not has_key():
        raise PrintfulError("PRINTFUL_API_KEY not set — add it to .env")
    url = f"{BASE}{path}"
    data = json.dumps(body).encode() if body is not None else None
    headers = {
        "Authorization": f"Bearer {api_key()}",
        "Accept": "application/json",
        "User-Agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) "
                      "AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0 Safari/537.36",
    }
    if data is not None:
        headers["Content-Type"] = "application/json"
    if store_id:
        headers["X-PF-Store-Id"] = str(store_id)
    req = urllib.request.Request(url, data=data, headers=headers, method=method)
    try:
        with urllib.request.urlopen(req, timeout=40, context=_SSL_CTX) as r:
            return json.loads(r.read() or "null")
    except urllib.error.HTTPError as e:
        detail = e.read().decode(errors="replace")[:500]
        raise PrintfulError(f"Printful {e.code} on {method} {path}: {detail}") from None
    except urllib.error.URLError as e:
        raise PrintfulError(f"Network error reaching Printful: {e.reason}") from None


# ------------------------------------------------------------------ catalog -- #
def list_catalog(category_id: int | None = None, limit: int = 100,
                 offset: int = 0) -> list:
    """Catalog products (the blanks Printful can print). No store needed."""
    q = f"?limit={limit}&offset={offset}"
    if category_id is not None:
        q += f"&category_id={category_id}"
    r = _request("GET", f"/products{q}")
    return r.get("result", [])


def categories() -> list:
    return _request("GET", "/categories").get("result", {}).get("categories", [])


def get_product(product_id: int | str) -> dict:
    """A catalog product + all its variants."""
    return _request("GET", f"/products/{product_id}").get("result", {})


def get_variant(variant_id: int | str) -> dict:
    """A single catalog variant (+ its parent product). Resolves an integer UID."""
    return _request("GET", f"/products/variant/{variant_id}").get("result", {})


def list_stores() -> list:
    """Stores on this account (empty until one is created in the dashboard)."""
    return _request("GET", "/stores").get("result", [])


# Loose map from Printful product type_name -> our garment category, so a
# Printful variant can be type-checked the same way as a Gelato one.
_TYPE_TO_CATEGORY = [
    ("hoodie", "hoodie"), ("sweatshirt", "sweatshirt"), ("crewneck", "sweatshirt"),
    ("tank", "tank-top"), ("polo", "polo"), ("t-shirt", "t-shirt"), ("tee", "t-shirt"),
]


def category_of(product: dict) -> str | None:
    name = (product.get("type_name") or product.get("title") or "").lower()
    for needle, cat in _TYPE_TO_CATEGORY:
        if needle in name:
            return cat
    return None


def verify_variant(variant_id: str, expected_category: str | None = None) -> tuple[bool, str]:
    """LIVE guard: the variant must resolve in Printful's catalog, and (when we
    can tell) its product type must match the expected garment category."""
    if not has_key():
        return True, "skipped (no key)"
    try:
        res = get_variant(variant_id)
    except PrintfulError as e:
        msg = str(e)
        if any(code in msg for code in (" 400", " 404", " 422")):
            return False, f"Printful rejected variant {variant_id}: {msg[:120]}"
        return True, f"live check skipped ({msg[:80]})"
    product = res.get("product", {}) or {}
    cat = category_of(product)
    if expected_category and cat and cat != expected_category:
        return False, (f"Type mismatch: Printful variant {variant_id} is "
                       f"'{cat}', expected '{expected_category}'")
    return True, "ok"

File: workers/src/lib/printful.ts
/**
 * Printful API client — SECOND fulfilment provider (apparel), ported 1:1 from
 * printful.py. Confirmed dormant in practice (every current catalog product
 * routes "local-sa") but ported faithfully in case it's ever activated.
 *
 * Auth: `Authorization: Bearer <PRINTFUL_API_KEY>` — a Workers secret
 * (`wrangler secret put PRINTFUL_API_KEY`), passed in via env, not a parsed
 * .env file (Workers has no filesystem / no need for the old dotenv shim).
 */

export interface PrintfulEnv {
  PRINTFUL_API_KEY?: string;
}

const BASE = "https://api.printful.com";

export class PrintfulError extends Error {}

export function hasKey(env: PrintfulEnv): boolean {
  return Boolean(env.PRINTFUL_API_KEY?.trim());
}

async function request(
  env: PrintfulEnv,
  method: string,
  path: string,
  body?: unknown,
  storeId?: string
): Promise<unknown> {
  if (!hasKey(env)) throw new PrintfulError("PRINTFUL_API_KEY not set");
  const headers: Record<string, string> = {
    Authorization: `Bearer ${env.PRINTFUL_API_KEY}`,
    Accept: "application/json",
    "User-Agent":
      "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0 Safari/537.36",
  };
  if (body !== undefined) headers["Content-Type"] = "application/json";
  if (storeId) headers["X-PF-Store-Id"] = storeId;

  let res: Response;
  try {
    res = await fetch(`${BASE}${path}`, {
      method,
      headers,
      body: body !== undefined ? JSON.stringify(body) : undefined,
    });
  } catch (e) {
    throw new PrintfulError(`Network error reaching Printful: ${(e as Error).message}`);
  }
  if (!res.ok) {
    const detail = (await res.text()).slice(0, 500);
    throw new PrintfulError(`Printful ${res.status} on ${method} ${path}: ${detail}`);
  }
  return res.json();
}

export async function getVariant(env: PrintfulEnv, variantId: string | number): Promise<any> {
  const r = (await request(env, "GET", `/products/variant/${variantId}`)) as { result?: unknown };
  return r.result ?? {};
}

const TYPE_TO_CATEGORY: [string, string][] = [
  ["hoodie", "hoodie"],
  ["sweatshirt", "sweatshirt"],
  ["crewneck", "sweatshirt"],
  ["tank", "tank-top"],
  ["polo", "polo"],
  ["t-shirt", "t-shirt"],
  ["tee", "t-shirt"],
];

export function categoryOf(product: { type_name?: string; title?: string }): string | null {
  const name = (product.type_name ?? product.title ?? "").toLowerCase();
  for (const [needle, cat] of TYPE_TO_CATEGORY) {
    if (name.includes(needle)) return cat;
  }
  return null;
}

/** LIVE guard: the variant must resolve in Printful's catalog, and (when we can
 * tell) its product type must match the expected garment category. */
export async function verifyVariant(
  env: PrintfulEnv,
  variantId: string,
  expectedCategory?: string | null
): Promise<[boolean, string]> {
  if (!hasKey(env)) return [true, "skipped (no key)"];
  let res: any;
  try {
    res = await getVariant(env, variantId);
  } catch (e) {
    const msg = (e as Error).message;
    if ([" 400", " 404", " 422"].some((code) => msg.includes(code))) {
      return [false, `Printful rejected variant ${variantId}: ${msg.slice(0, 120)}`];
    }
    return [true, `live check skipped (${msg.slice(0, 80)})`];
  }
  const product = res.product ?? {};
  const cat = categoryOf(product);
  if (expectedCategory && cat && cat !== expectedCategory) {
    return [false, `Type mismatch: Printful variant ${variantId} is '${cat}', expected '${expectedCategory}'`];
  }
  return [true, "ok"];
}
